### PR TITLE
add disk io metric

### PR DESCRIPTION
**Pull request by andyxning**
Address #690.

This PR will add disk io metrics to heapster. It will add these metrics:
* `disk/io_read_bytes`: Number of bytes read from a disk partition
* `disk/io_write_bytes`: Number of bytes written to a disk partition
* `disk/io_read_bytes_rate`: Number of bytes read from a disk partition per second
*  `disk/io_write_bytes_rate`: Number of bytes written to a disk partition per second

all these metrics will add `resource_id` label with values in `major:minor` format.

FYI：[kernel blkio cgroup](https://www.kernel.org/doc/Documentation/cgroup-v1/blkio-controller.txt)

### CODE REVIEW

**Comment by andyxning** [comment]
@piosz @DirectXMan12 @mwielgus PTAL.

**Comment by piosz** [comment]
@timstclair @vishh can one of you take a look? Does it make sense to have those metrics?

**Comment by vishh** [comment]
In on-prem clusters, disk is often a contended resource. We do not provide
any isolation for disk yet. The best we can do is to expose usage metrics.
Disk IO is a common issue and exposing that metrics to users will enable
them to identify offending pods and take action. So I'd recommend adding
metrics for "space", "inodes" and "IO" for disk

On Fri, Jan 13, 2017 at 12:48 AM, Piotr Szczesniak <notifications@github.com

**Comment by andyxning** [comment]
@vishh 

> So I'd recommend adding metrics for "space", "inodes" and "IO" for disk

* There is already exists `filesystem/usage`, `filesystemd/limit` and `filesystem/available` for `space`. 
* As for `inodes`, it seems that `blkio` cgroup has no info about this. I have found that inodes info is available in [cAdvisor FsStats](https://github.com/google/cadvisor/blob/master/info/v1/container.go#L436-L443). Will add this in another PR. :)
* `blkio` cgroup supports setting resource Throttling/Upper limit policy for `IO`. This is what this PR does.

**Comment by andyxning** [comment]
Comments addressed. Friendly ping @timstclair @vishh

**Comment by andyxning** [comment]
Friendly ping again. :) @timstclair @vishh

**Comment by piosz** [comment]
Removing out of 1.3

**Comment by dnavre** [comment]
This PR will be very useful especially if one runs database shards in kubernetes. Please do merge it :)

**Comment by weikinhuang** [comment]
Has there been any updates on this? When running a self hosted cluster with spindle disks this is very useful information to have.

**Comment by DirectXMan12** [comment]
it needs to be rebased before it can be merged.

**Comment by jingxu97** [comment]
@andyxning, will you pick up this PR again?

**Comment by andyxning** [comment]
@jingxu97 Yes. We need this also. I will rebase and resolve the conflicts in next week. 

BTW, any thoughts on the implementations or suggestions.

**Comment by andyxning** [comment]
@DirectXMan12 @piosz @loburm @vishh @dnavre @weikinhuang @jingxu97 This PR is ready to review. :) 

PTAL.

**Comment by andyxning** [comment]
Ping @DirectXMan12 @piosz @loburm since #1879 has been merge. This PR is ready to review and should be ready to merge too. PTAL.

**Comment by andyxning** [comment]
Ping @kubernetes/heapster-maintainers 

Sorry for the complain. But can someone help me to review this PR instead of just reviewing and merging Google related ones. :(

**Comment by tallclair** [comment]
This only adds the metrics for the cAdvisor source, but we're moving towards the summary source instead (parsed here: https://github.com/kubernetes/heapster/blob/master/metrics/sources/summary/summary.go).

@piosz What is the future plan for the cAdvisor source? We may remove the cAdvisor stats from the Kubelet at some point in the future.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
sorry for the delay.  This needs to add metrics from the summary source, since that's the reccomended source these days.

**Comment by andyxning** [comment]
> sorry for the delay. This needs to add metrics from the summary source, since that's the reccomended source these days.

According to https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/stats/v1alpha1/types.go#L110-L135, the disk io related data has not been added to stats api. I prefer to add `legacy` stats api support for now. Adding summary api support is another long term plan just like network metrics #1788 .

**Comment by andyxning** [comment]
Ping @DirectXMan12

**Comment by DirectXMan12** [comment]
I'd *very* much prefer not to add anything new to the legacy source, *especially* if it's not in the summary API.  Please get it added to the summary API.

**Comment by andyxning** [comment]
Ping @piosz

**Comment by piosz** [comment]
@andyxning apologies for the delay. I'm fine with this PR. As @DirectXMan12 wrote please have in mind that `/stats` endpoint is deprecated in Kubelet and will be removed from there at some point, especially Heapster while using it consumes much more resources.

At the same time it's not deprecated in pure Heapster to some degree - we just need to tune Heapster a bit to work with standalone cadvisor.

**Comment by andyxning** [comment]
Thanks @DirectXMan12 @piosz . 

Yes, it is quite a question when the cluster becomes more and more large. With the original cAdvisor `stats` endpoint, more and more redundant data needs to be scraped as the cluster grows. This is quite a huge overhead. 

Btw, the summary api needs to be enhanced to accomplish our basic metrics requirements. :) Will do some works on that. :)

**Comment by weikinhuang** [comment]
How do I access this in `v1.5.0` within influxdb/grafana? When I try to add a new dashboard graph, that field seems to be missing:

screen shot 2017-12-13 at 12 41 36 pm

**Comment by andyxning** [comment]
@weikinhuang 
```
disk/io_read_bytes_rate: Number of bytes read from a disk partition per second
disk/io_write_bytes_rate: Number of bytes written to a disk partition per second
```

**Comment by weikinhuang** [comment]
I don't see them in the recorded metrics

**Comment by andyxning** [comment]
@weikinhuang How is your configuration for `--source`?

**Comment by weikinhuang** [comment]
@andyxning Here is my deployment spec for heapster (mostly from the kubernetes/addons dir)
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: heapster-v1.5.0
  namespace: kube-system
  labels:
    k8s-app: heapster
    kubernetes.io/cluster-service: "true"
    addonmanager.kubernetes.io/mode: Reconcile
    version: v1.5.0
spec:
  replicas: 1
  selector:
    matchLabels:
      k8s-app: heapster
      version: v1.5.0
  template:
    metadata:
      labels:
        k8s-app: heapster
        version: v1.5.0
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
    spec:
      containers:
        - image: gcr.io/google_containers/heapster-amd64:v1.5.0
          name: heapster
          livenessProbe:
            httpGet:
              path: /healthz
              port: 8082
              scheme: HTTP
            initialDelaySeconds: 180
            timeoutSeconds: 5
          command:
            - /heapster
            - --source=kubernetes.summary_api:''
            - --sink=influxdb:http://monitoring-influxdb:8086
        - image: gcr.io/google_containers/heapster-amd64:v1.5.0
          name: eventer
          command:
            - /eventer
            - --source=kubernetes:''
            - --sink=influxdb:http://monitoring-influxdb:8086
        - image: gcr.io/google_containers/addon-resizer:1.7
          name: heapster-nanny
          resources:
            limits:
              cpu: 50m
              memory: 90Mi
            requests:
              cpu: 50m
              memory: 90Mi
          env:
            - name: MY_POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: MY_POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
          command:
            - /pod_nanny
            - --cpu=80m
            - --extra-cpu=0.5m
            - --memory=140Mi
            - --extra-memory=4Mi
            - --deployment=heapster-v1.5.0
            - --container=heapster
            - --poll-period=300000
        - image: gcr.io/google_containers/addon-resizer:1.7
          name: eventer-nanny
          resources:
            limits:
              cpu: 50m
              memory: 200Mi
            requests:
              cpu: 50m
              memory: 200Mi
          env:
            - name: MY_POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: MY_POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
          command:
            - /pod_nanny
            - --cpu=100m
            - --extra-cpu=0m
            - --memory=190Mi
            - --extra-memory=500Ki
            - --deployment=heapster-v1.5.0
            - --container=eventer
            - --poll-period=300000
      serviceAccountName: heapster
      tolerations:
        - key: "CriticalAddonsOnly"
          operator: "Exists"
```

**Comment by weikinhuang** [comment]
Ah. It looks like I'm using the summary API. Which does not have these metrics. Is there any issue or pr I can subscribe to for disk metrics in the summary API? Thanks @andyxning

**Comment by andyxning** [comment]
@weikinhuang As for now there is not pr for adding disk io metric in summary api. I am planning to add this in next week. If you can help and make a PR about this, i will be ok to review it. :)

**Comment by weikinhuang** [comment]
I would love to help, but my knowledge of golang's syntax is limited.

**Comment by andyxning** [comment]
@weikinhuang To that situation, i suggest to wait for another week maybe for this feature added to summary api source. I will give it a try.

**Comment by weikinhuang** [comment]
No problem, thanks @andyxning!

**Comment by Gimi** [comment]
@andyxning just wondering that are you still working on adding disk metrics to summary api source?

**Comment by Gimi** [comment]
I use "kubernetes" source and still am not seeing any disk metrics, any idea? I'm using gcr.io/google_containers/heapster:v1.5.0 running on kubernetes 1.8.6.

**Comment by andyxning** [comment]
Any logs for heapster?

**Comment by andyxning** [comment]
> just wondering that are you still working on adding disk metrics to summary api source?

Yes, i will work on this. But no time recently available working on this. :( 
Maybe one week later.

**Comment by Gimi** [comment]
@andyxning no error nor warning message from heapster's log.

**Comment by andyxning** [comment]
@Gimi Which sink backend do you use?

**Comment by Gimi** [comment]
@andyxning sorry for the late response, I use statsd.

**Comment by andyxning** [comment]
@Gimi Just have thought that do you use the summary data source? The disk io is currently only available on legacy data source.

**Comment by andyxning** [comment]
@Gimi I have test master branch with statsd sink and legacy data source and the disk io metrics is available like:
> node.192_168_0_1.batch.sandbox;beta_kubernetes_io/arch.disk/io_read_bytes_rate./dev/sdb:15381.686|g
node.192_168_0_1.batch.sandbox;beta_kubernetes_io/arch.disk/io_read_bytes_rate./dev/sda:-1.3200837e+09|g
node.192_168_0_1.batch.sandbox;beta_kubernetes_io/arch.disk/io_write_bytes_rate./dev/sdb:2.1783398e+06|g
node.192_168_0_1.batch.sandbox;beta_kubernetes_io/arch.disk/io_write_bytes_rate./dev/sda:-5.901337e+10|g

**Comment by Gimi** [comment]
thanks @andyxning for checking. I at the beginning did use the summary api resource, but I switched back to the legacy one when I found this ticket. But still did not see any disk metrics. I'll check again and see if I missed anything. Thank you.

**Pull request merged**
